File: utils/agents/AI_Agent_Code_Executor/extract_sql_queries_tool.py
"""n8n Code node that extracts SQL SELECT statements using strict JSON-emitter mode rules."""
from __future__ import annotations

import re
from typing import Any, Dict, List, Optional

def numeric_order(key: str) -> int:
    digits = "".join(ch for ch in str(key) if ch.isdigit())
    return int(digits) if digits else 10**6

def strip_sql_comments(sql: str) -> str:
    """Remove SQL comments from the query."""
    # Remove /* */ block comments
    sql = re.sub(r'/\*.*?\*/', '', sql, flags=re.DOTALL)
    # Remove -- line comments
    sql = re.sub(r'--.*$', '', sql, flags=re.MULTILINE)
    return sql

def is_read_only_query(sql: str) -> bool:
    """Check if the SQL query is read-only (SELECT or WITH...SELECT)."""
    sql_clean = strip_sql_comments(sql).strip().upper()
    
    # Check for data-changing keywords outside quoted strings
    forbidden_keywords = [
        'INSERT', 'UPDATE', 'DELETE', 'MERGE', 'CREATE', 'ALTER', 'DROP', 
        'TRUNCATE', 'GRANT', 'REVOKE', 'COMMENT', 'COPY'
    ]
    
    # Simple check for quoted strings and forbidden keywords
    # This is a basic implementation - for production, use a proper SQL parser
    for keyword in forbidden_keywords:
        if keyword in sql_clean:
            return False
    
    # Check for "SELECT ... INTO" pattern
    if 'SELECT' in sql_clean and 'INTO' in sql_clean:
        return False
    
    # Must start with SELECT or WITH
    return sql_clean.startswith(('SELECT', 'WITH'))

def extract_from_fenced_sql(text: str) -> Optional[str]:
    """Extract SQL from fenced code blocks with sql language tag."""
    pattern = r'```sql\s*\n(.*?)\n```'
    match = re.search(pattern, text, re.IGNORECASE | re.DOTALL)
    if match:
        sql = match.group(1).strip()
        if sql.endswith(';'):
            sql = sql[:-1]
        return sql
    return None

def extract_from_inline_sql(text: str) -> Optional[str]:
    """Extract SQL from single-backtick inline snippets."""
    pattern = r'`([^`]+)`'
    match = re.search(pattern, text)
    if match:
        sql = match.group(1).strip()
        if sql.endswith(';'):
            sql = sql[:-1]
        return sql
    return None

def extract_from_statement_scan(text: str) -> Optional[str]:
    """Extract SQL using statement scan (WITH...SELECT or SELECT)."""
    text_clean = text.strip()
    
    # Check for WITH...SELECT pattern
    if text_clean.upper().startswith('WITH'):
        # Find the end of the statement (first semicolon not in quotes)
        # This is a simplified approach - for production, use proper SQL parsing
        end_pos = text_clean.find(';')
        if end_pos != -1:
            sql = text_clean[:end_pos].strip()
            if sql.endswith(';'):
                sql = sql[:-1]
            return sql
        return text_clean
    
    # Check for SELECT pattern
    select_pos = text_clean.upper().find('SELECT')
    if select_pos != -1:
        sql = text_clean[select_pos:].strip()
        end_pos = sql.find(';')
        if end_pos != -1:
            sql = sql[:end_pos].strip()
        if sql.endswith(';'):
            sql = sql[:-1]
        return sql
    
    return None

def apply_identifier_correction(sql: str) -> str:
    """Apply the only allowed identifier correction: public.user -> public."user"."""
    return sql.replace('public.user', 'public."user"')

def extract_sql_from_text(text: str) -> Optional[str]:
    """Extract SQL from text using strict precedence rules."""
    if not isinstance(text, str) or not text.strip():
        return None
    
    # Precedence 1: Fenced SQL with language tag
    sql = extract_from_fenced_sql(text)
    if sql and is_read_only_query(sql):
        return apply_identifier_correction(sql)
    
    # Precedence 2: Single-backtick inline SQL
    sql = extract_from_inline_sql(text)
    if sql and is_read_only_query(sql):
        return apply_identifier_correction(sql)
    
    # Precedence 3: Statement scan
    sql = extract_from_statement_scan(text)
    if sql and is_read_only_query(sql):
        return apply_identifier_correction(sql)
    
    return None

def validate_schema_tables(sql: str, schema_text: Optional[str]) -> bool:
    """Validate that referenced tables exist in the schema."""
    if not schema_text:
        return True
    
    # Extract table names from FROM and JOIN clauses
    # This is a simplified approach - for production, use proper SQL parsing
    table_pattern = r'(?:FROM|JOIN)\s+([a-zA-Z_][a-zA-Z0-9_.]*)'
    matches = re.findall(table_pattern, sql, re.IGNORECASE)
    
    # Check if tables exist in schema
    for table in matches:
        if table not in schema_text:
            return False
    
    return True

def extract_queries_strict(input_items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Extracts SQL queries using strict JSON-emitter mode rules."""
    collected: List[Dict[str, Any]] = []
    
    for item in input_items:
        payload = item.get("json", {}) if isinstance(item, dict) else {}
        
        # Look for "response" array first (primary input format)
        response_array = payload.get("response")
        if isinstance(response_array, list):
            schema_text = payload.get("schema")
            for text in response_array:
                sql = extract_sql_from_text(text)
                if sql and validate_schema_tables(sql, schema_text):
                    collected.append({
                        "query": sql,
                        "params": []
                    })
            continue
        
        # Fallback to legacy formats for backward compatibility
        queries_array = payload.get("queries")
        if isinstance(queries_array, list):
            for entry in queries_array:
                text = (entry or {}).get("query", "")
                sql = extract_sql_from_text(text)
                if sql:
                    collected.append({
                        "query": sql,
                        "params": entry.get("params", []),
                    })
            continue

        data_section = payload.get("data")
        if isinstance(data_section, dict):
            for key in sorted(data_section.keys(), key=numeric_order):
                entry = data_section.get(key) or {}
                text = entry.get("query", "")
                sql = extract_sql_from_text(text)
                if sql:
                    collected.append({
                        "query": sql,
                        "params": entry.get("params", []),
                    })
            continue

        query_keys = [key for key in payload.keys() if key.lower().startswith("query")]
        for key in sorted(query_keys, key=numeric_order):
            text = payload.get(key, "")
            sql = extract_sql_from_text(text)
            if sql:
                collected.append({"query": sql, "params": []})

    return collected

def main(input_data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Entry point invoked by n8n when the AI agent calls this tool."""
    tool_args = input_data[0].get("json", {})
    input_items_for_extraction = tool_args.get("input_items")

    if not input_items_for_extraction:
        return [{"json": {"error": "Missing 'input_items' for query extraction."}}]

    extracted_queries = extract_queries_strict(input_items_for_extraction)
    return [{"json": {"extracted_queries": extracted_queries}}]

File: utils/agents/AI_Agent_Code_Executor/extract_sql_queries_tool.js
/**
 * n8n Code node that extracts SQL SELECT statements using strict JSON-emitter mode rules.
 */

/**
 * Helper function to sort keys numerically
 */
function numericOrder(key) {
  const digits = key.replace(/\D/g, '');
  return digits ? parseInt(digits, 10) : 1000000;
}

/**
 * Remove SQL comments from the query.
 */
function stripSqlComments(sql) {
  return sql
    .replace(/\/\*[\s\S]*?\*\//g, '') // Remove /* */ block comments
    .replace(/--.*$/gm, ''); // Remove -- line comments
}

/**
 * Check if the SQL query is read-only (SELECT or WITH...SELECT).
 */
function isReadOnlyQuery(sql) {
  const sqlClean = stripSqlComments(sql).trim().toUpperCase();
  
  // Check for data-changing keywords outside quoted strings
  const forbiddenKeywords = [
    'INSERT', 'UPDATE', 'DELETE', 'MERGE', 'CREATE', 'ALTER', 'DROP',
    'TRUNCATE', 'GRANT', 'REVOKE', 'COMMENT', 'COPY'
  ];
  
  // Simple check for quoted strings and forbidden keywords
  // This is a basic implementation - for production, use a proper SQL parser
  for (const keyword of forbiddenKeywords) {
    if (sqlClean.includes(keyword)) {
      return false;
    }
  }
  
  // Check for "SELECT ... INTO" pattern
  if (sqlClean.includes('SELECT') && sqlClean.includes('INTO')) {
    return false;
  }
  
  // Must start with SELECT or WITH
  return sqlClean.startsWith('SELECT') || sqlClean.startsWith('WITH');
}

/**
 * Extract SQL from fenced code blocks with sql language tag.
 */
function extractFromFencedSql(text) {
  const pattern = /```sql\s*\n(.*?)\n```/is;
  const match = text.match(pattern);
  if (match) {
    let sql = match[1].trim();
    if (sql.endsWith(';')) {
      sql = sql.slice(0, -1);
    }
    return sql;
  }
  return null;
}

/**
 * Extract SQL from single-backtick inline snippets.
 */
function extractFromInlineSql(text) {
  const pattern = /`([^`]+)`/;
  const match = text.match(pattern);
  if (match) {
    let sql = match[1].trim();
    if (sql.endsWith(';')) {
      sql = sql.slice(0, -1);
    }
    return sql;
  }
  return null;
}

/**
 * Extract SQL using statement scan (WITH...SELECT or SELECT).
 */
function extractFromStatementScan(text) {
  const textClean = text.trim();
  
  // Check for WITH...SELECT pattern
  if (textClean.toUpperCase().startsWith('WITH')) {
    const endPos = textClean.indexOf(';');
    if (endPos !== -1) {
      let sql = textClean.slice(0, endPos).trim();
      if (sql.endsWith(';')) {
        sql = sql.slice(0, -1);
      }
      return sql;
    }
    return textClean;
  }
  
  // Check for SELECT pattern
  const selectPos = textClean.toUpperCase().indexOf('SELECT');
  if (selectPos !== -1) {
    let sql = textClean.slice(selectPos).trim();
    const endPos = sql.indexOf(';');
    if (endPos !== -1) {
      sql = sql.slice(0, endPos).trim();
    }
    if (sql.endsWith(';')) {
      sql = sql.slice(0, -1);
    }
    return sql;
  }
  
  return null;
}

/**
 * Apply the only allowed identifier correction: public.user -> public."user".
 */
function applyIdentifierCorrection(sql) {
  return sql.replace(/public\.user/g, 'public."user"');
}

/**
 * Extract SQL from text using strict precedence rules.
 */
function extractSqlFromText(text) {
  if (!text || !text.trim()) {
    return null;
  }
  
  // Precedence 1: Fenced SQL with language tag
  let sql = extractFromFencedSql(text);
  if (sql && isReadOnlyQuery(sql)) {
    return applyIdentifierCorrection(sql);
  }
  
  // Precedence 2: Single-backtick inline SQL
  sql = extractFromInlineSql(text);
  if (sql && isReadOnlyQuery(sql)) {
    return applyIdentifierCorrection(sql);
  }
  
  // Precedence 3: Statement scan
  sql = extractFromStatementScan(text);
  if (sql && isReadOnlyQuery(sql)) {
    return applyIdentifierCorrection(sql);
  }
  
  return null;
}

/**
 * Validate that referenced tables exist in the schema.
 */
function validateSchemaTables(sql, schemaText) {
  if (!schemaText) {
    return true;
  }
  
  // Extract table names from FROM and JOIN clauses
  // This is a simplified approach - for production, use proper SQL parsing
  const tablePattern = /(?:FROM|JOIN)\s+([a-zA-Z_][a-zA-Z0-9_.]*)/gi;
  const matches = sql.match(tablePattern);
  
  if (matches) {
    for (const match of matches) {
      const tableName = match.replace(/(?:FROM|JOIN)\s+/i, '');
      if (!schemaText.includes(tableName)) {
        return false;
      }
    }
  }
  
  return true;
}

/**
 * Extracts SQL queries using strict JSON-emitter mode rules.
 */
function extractQueriesStrict(inputItems) {
  const collected = [];
  
  for (const item of inputItems) {
    const payload = item.json || {};
    
    // Look for "response" array first (primary input format)
    const responseArray = payload.response;
    if (Array.isArray(responseArray)) {
      const schemaText = payload.schema;
      for (const text of responseArray) {
        const sql = extractSqlFromText(text);
        if (sql && validateSchemaTables(sql, schemaText)) {
          collected.push({
            query: sql,
            params: []
          });
        }
      }
      continue;
    }
    
    // Fallback to legacy formats for backward compatibility
    const queriesArray = payload.queries;
    if (Array.isArray(queriesArray)) {
      for (const entry of queriesArray) {
        const text = entry?.query || '';
        const sql = extractSqlFromText(text);
        if (sql) {
          collected.push({
            query: sql,
            params: entry.params || []
          });
        }
      }
      continue;
    }

    const dataSection = payload.data;
    if (typeof dataSection === 'object' && dataSection !== null) {
      const sortedKeys = Object.keys(dataSection).sort((a, b) => numericOrder(a) - numericOrder(b));
      for (const key of sortedKeys) {
        const entry = dataSection[key] || {};
        const text = entry.query || '';
        const sql = extractSqlFromText(text);
        if (sql) {
          collected.push({
            query: sql,
            params: entry.params || []
          });
        }
      }
      continue;
    }

    const queryKeys = Object.keys(payload).filter(key => key.toLowerCase().startsWith('query'));
    const sortedQueryKeys = queryKeys.sort((a, b) => numericOrder(a) - numericOrder(b));
    for (const key of sortedQueryKeys) {
      const text = payload[key] || '';
      const sql = extractSqlFromText(text);
      if (sql) {
        collected.push({ query: sql, params: [] });
      }
    }
  }

  return collected;
}

/**
 * Entry point invoked by n8n when the AI agent calls this tool.
 */
async function main(inputData) {
  try {
    const toolArgs = inputData[0].json || {};
    const inputItemsForExtraction = toolArgs.input_items;

    if (!inputItemsForExtraction) {
      return [{ json: { error: "Missing 'input_items' for query extraction." } }];
    }

    const extractedQueries = extractQueriesStrict(inputItemsForExtraction);
    return [{ json: { extracted_queries: extractedQueries } }];
  } catch (error) {
    return [{ json: { error: error.message } }];
  }
}

module.exports = { main };
